File: internal/agent/agent.go
// Package agent is the outbound-only popfleet agent: dial the broker,
// heartbeat, spawn one PTY per session id, relay bytes. It opens no
// listening sockets and reconnects forever.
package agent

import (
	"fmt"
	"log"
	"math/rand/v2"
	"net/url"
	"os"
	"os/exec"
	"strings"
	"sync"
	"syscall"
	"time"

	"github.com/creack/pty"
	"github.com/gorilla/websocket"

	"github.com/ejoliet/popfleet/internal/e2e"
	"github.com/ejoliet/popfleet/internal/proto"
)

// Version is stamped at release time via
// -ldflags="-X github.com/ejoliet/popfleet/internal/agent.Version=<tag>".
// Plain `go build` keeps "dev" so unstamped binaries are honest about it.
var Version = "dev"

// Run dials the broker and never returns (reconnect loop with backoff
// 1 s doubling to 30 s, ±20% jitter), except on a config error.
// e2eKey nil = plaintext v1 (LAN Go broker); 32 bytes = protocol v1e,
// required by the Worker relay.
func Run(rawURL, token, name string, e2eKey []byte) error {
	wsURL, err := agentWSURL(rawURL)
	if err != nil {
		return err
	}
	backoff := time.Second
	for {
		ok, err := session(wsURL, token, name, e2eKey)
		if err != nil {
			log.Printf("agent: connection lost: %v", err)
		}
		if ok {
			backoff = time.Second // successful hello resets backoff
		}
		jitter := 0.8 + rand.Float64()*0.4 // ±20%
		time.Sleep(time.Duration(float64(backoff) * jitter))
		backoff = nextBackoff(backoff)
	}
}

// nextBackoff doubles up to the 30 s cap. Split out of Run so the sequence is
// testable: Run's loop never returns, so it has no seam of its own.
func nextBackoff(d time.Duration) time.Duration {
	if d *= 2; d > maxBackoff {
		return maxBackoff
	}
	return d
}

const maxBackoff = 30 * time.Second

// agentWSURL turns http(s):// or ws(s):// into the /ws/agent endpoint.
func agentWSURL(raw string) (string, error) {
	u, err := url.Parse(raw)
	if err != nil {
		return "", fmt.Errorf("bad POPFLEET_URL: %w", err)
	}
	switch u.Scheme {
	case "http":
		u.Scheme = "ws"
	case "https":
		u.Scheme = "wss"
	case "ws", "wss":
	default:
		return "", fmt.Errorf("bad POPFLEET_URL scheme %q (want http/https/ws/wss)", u.Scheme)
	}
	u.Path = strings.TrimRight(u.Path, "/") + "/ws/agent"
	return u.String(), nil
}

type ptySession struct {
	f   *os.File
	cmd *exec.Cmd
	enc *e2e.Session // nil in plaintext v1 mode
}

// session runs one connection to the broker. Returns hello-succeeded plus
// the terminal error. On any exit every PTY is killed.
func session(wsURL, token, name string, e2eKey []byte) (helloOK bool, err error) {
	ws, _, err := websocket.DefaultDialer.Dial(wsURL, nil)
	if err != nil {
		return false, err
	}
	defer ws.Close()

	var wmu sync.Mutex // gorilla allows one concurrent writer
	send := func(m proto.Msg) error {
		wmu.Lock()
		defer wmu.Unlock()
		return ws.WriteJSON(m)
	}
	// hb is sent as this exact byte string, not via WriteJSON: the Worker
	// relay absorbs it with a WebSocket auto-response pair (hibernation stays
	// cheap) and that match is byte-exact — Encode's trailing newline or a
	// reordered field would silently wake the DO on every heartbeat.
	sendHB := func() error {
		wmu.Lock()
		defer wmu.Unlock()
		return ws.WriteMessage(websocket.TextMessage, []byte(`{"t":"hb"}`))
	}

	if err := send(proto.Msg{T: "hello", Token: token, Name: name, Ver: Version, E2E: len(e2eKey) > 0}); err != nil {
		return false, err
	}
	ws.SetReadDeadline(time.Now().Add(10 * time.Second))
	var m proto.Msg
	if err := ws.ReadJSON(&m); err != nil || m.T != "hello_ok" {
		if err == nil && m.T == "err" && m.Msg != "" {
			// v2 relay rejects with a reason (e.g. e2e required); surface it.
			return false, fmt.Errorf("hello rejected: %s", m.Msg)
		}
		return false, fmt.Errorf("hello rejected (bad or revoked token?): %v", err)
	}
	log.Printf("agent: connected as machine %s", m.ID)

	var pmu sync.Mutex
	ptys := map[string]*ptySession{}
	defer func() { // socket loss kills every PTY (requirement 7)
		pmu.Lock()
		for sid, p := range ptys {
			p.cmd.Process.Kill()
			p.f.Close()
			delete(ptys, sid)
		}
		pmu.Unlock()
	}()

	// Liveness: broker sends nothing to an idle agent, so JSON reads alone
	// can't detect a dead broker. Ping alongside each hb; pongs (gorilla
	// auto-replies to pings) extend the read deadline.
	const liveness = 45 * time.Second
	ws.SetReadDeadline(time.Now().Add(liveness))
	ws.SetPongHandler(func(string) error {
		return ws.SetReadDeadline(time.Now().Add(liveness))
	})

	done := make(chan struct{})
	defer close(done)
	go func() { // heartbeat every 10 s
		t := time.NewTicker(10 * time.Second)
		defer t.Stop()
		for {
			select {
			case <-done:
				return
			case <-t.C:
				if sendHB() != nil {
					return
				}
				ws.WriteControl(websocket.PingMessage, nil, time.Now().Add(5*time.Second))
			}
		}
	}()

	for {
		var m proto.Msg
		if err := ws.ReadJSON(&m); err != nil {
			return true, err
		}
		switch m.T {
		case "open":
			pmu.Lock()
			if _, exists := ptys[m.Sid]; exists { // duplicate open ignored
				pmu.Unlock()
				continue
			}
			pmu.Unlock()
			var enc *e2e.Session
			cmdline := m.Cmd
			if len(e2eKey) > 0 {
				var b []byte
				var err error
				if enc, err = e2e.NewSession(e2eKey, m.Sid); err == nil && cmdline != "" {
					if b, err = enc.Open(cmdline); err == nil {
						cmdline = string(b)
					}
				}
				if err != nil { // GCM fail = never run what we cannot authenticate
					log.Printf("agent: e2e open rejected for sid %s: %v", m.Sid, err)
					send(proto.Msg{T: "err", Sid: m.Sid, Msg: "e2e decrypt failed (key mismatch or tampered cmd)"})
					continue
				}
			}
			p, err := spawn(cmdline)
			if err != nil {
				log.Printf("agent: spawn failed for sid %s: %v", m.Sid, err)
				send(proto.Msg{T: "exit", Sid: m.Sid, Code: proto.Int(127)})
				continue
			}
			p.enc = enc
			pmu.Lock()
			ptys[m.Sid] = p
			pmu.Unlock()
			go pump(p, m.Sid, send, func() {
				pmu.Lock()
				delete(ptys, m.Sid)
				pmu.Unlock()
			})
		case "in":
			pmu.Lock()
			p, ok := ptys[m.Sid]
			pmu.Unlock()
			if !ok { // unknown sid: ignored, never an error
				continue
			}
			var b []byte
			var err error
			if p.enc != nil {
				b, err = p.enc.Open(m.Data)
			} else {
				b, err = proto.Dec(m.Data)
			}
			if err != nil {
				if p.enc != nil { // tampered/forged keystrokes: kill the session
					log.Printf("agent: e2e in rejected for sid %s: %v", m.Sid, err)
					send(proto.Msg{T: "err", Sid: m.Sid, Msg: "e2e decrypt failed (tampered frame)"})
					p.cmd.Process.Kill() // pump sends exit and forgets the sid
				}
				continue
			}
			p.f.Write(b)
		case "resize":
			pmu.Lock()
			if p, ok := ptys[m.Sid]; ok {
				pty.Setsize(p.f, &pty.Winsize{Cols: uint16(m.C), Rows: uint16(m.R)})
			}
			pmu.Unlock()
		case "close":
			pmu.Lock()
			if p, ok := ptys[m.Sid]; ok {
				p.cmd.Process.Kill() // pump sends exit and forgets the sid
			}
			pmu.Unlock()
		}
	}
}

// spawn starts a login shell (or `$SHELL -c cmd`) on a fresh PTY.
func spawn(cmdline string) (*ptySession, error) {
	shell := os.Getenv("SHELL")
	if shell == "" {
		shell = "/bin/sh"
	}
	var c *exec.Cmd
	if cmdline != "" {
		c = exec.Command(shell, "-c", cmdline)
	} else {
		c = exec.Command(shell, "-l")
	}
	c.Env = append(os.Environ(), "TERM=xterm-256color")
	f, err := pty.Start(c)
	if err != nil {
		return nil, err
	}
	return &ptySession{f: f, cmd: c}, nil
}

// Output coalescing: flush every 16 ms or 4 KiB, whichever first. A `yes`
// firehose as frame-per-read is a per-frame cost through the Worker relay
// (request count, duration); batched it is a non-issue. Agent-side only,
// no protocol impact, and the v1 Go broker benefits the same way.
const (
	flushEvery = 16 * time.Millisecond
	flushBytes = 4 * 1024
)

// pump copies PTY output to the broker (coalesced, encrypted when the
// session has an e2e key), then reports exit and forgets the sid.
func pump(p *ptySession, sid string, send func(proto.Msg) error, forget func()) {
	chunks := make(chan []byte, 32)
	go func() { // reader: PTY -> chunks; closes on shell exit
		defer close(chunks)
		for {
			buf := make([]byte, 8192)
			n, err := p.f.Read(buf)
			if n > 0 {
				chunks <- buf[:n]
			}
			if err != nil {
				return // EIO on shell exit is the normal path
			}
		}
	}()

	encode := proto.Enc
	if p.enc != nil {
		encode = p.enc.Seal
	}
	var buf []byte
	var deadline <-chan time.Time
	flush := func() bool {
		if len(buf) == 0 {
			return true
		}
		ok := send(proto.Msg{T: "out", Sid: sid, Data: encode(buf)}) == nil
		buf, deadline = nil, nil
		return ok
	}
loop:
	for {
		select {
		case c, ok := <-chunks:
			if !ok {
				flush()
				break loop
			}
			buf = append(buf, c...)
			if len(buf) >= flushBytes {
				if !flush() {
					break loop // socket gone; session() defer kills the PTY
				}
			} else if deadline == nil {
				deadline = time.After(flushEvery)
			}
		case <-deadline:
			if !flush() {
				break loop
			}
		}
	}
	// Unblock the reader if it is still mid-send (send-failure path); it
	// exits once session()'s defer kills the PTY and the read errors.
	go func() {
		for range chunks {
		}
	}()
	code := 0
	if err := p.cmd.Wait(); err != nil {
		if ee, ok := err.(*exec.ExitError); ok {
			code = ee.ExitCode()
			// ExitCode() is -1 when the shell died on a signal; the protocol
			// says 128+N (137 for SIGKILL), matching the Python agent.
			if ws, ok := ee.Sys().(syscall.WaitStatus); ok && ws.Signaled() {
				code = 128 + int(ws.Signal())
			}
		} else {
			code = 1
		}
	}
	p.f.Close()
	forget()
	send(proto.Msg{T: "exit", Sid: sid, Code: proto.Int(code)})
}

File: contrib/agent.py
#!/usr/bin/env python3
"""popfleet reference agent: dials out, holds one WebSocket, serves PTYs.

Embed it in your own app (config is env-only, never argv):

    import os, threading, agent
    os.environ["POPFLEET_URL"] = "https://fleet.example"     # broker base URL
    os.environ["POPFLEET_TOKEN"] = "<enrollment token from the panel>"
    os.environ["POPFLEET_NAME"] = "my-app"        # optional, defaults to hostname
    os.environ["POPFLEET_E2E_KEY"] = "<base64>"   # required for the Worker relay
    threading.Thread(target=agent.main, daemon=True).start()

Standalone: POPFLEET_URL=... POPFLEET_TOKEN=... python3 agent.py
Requires:   Python 3.9+ and `pip install websockets`; with POPFLEET_E2E_KEY
            set, also `pip install cryptography` (AES-256-GCM).
Protocol:   docs/PROTOCOL.md v1; with POPFLEET_E2E_KEY, the v1e payload
            encryption from docs/RDD-v2.md on top (frame shapes unchanged).
"""

import asyncio
import base64
import errno
import fcntl
import hashlib
import hmac
import json
import os
import pty
import random
import signal
import socket
import struct
import sys
import termios
from urllib.parse import urlsplit, urlunsplit

import websockets

VER = "2.0.0"
HEARTBEAT_S = 10
BACKOFF_START_S = 1.0
BACKOFF_MAX_S = 30.0
READ_SIZE = 65536
FLUSH_S = 0.016          # coalesce PTY output: flush every 16 ms ...
FLUSH_BYTES = 4096       # ... or 4 KiB, whichever comes first
# hb is this exact byte string: the Worker relay absorbs it with a
# byte-exact WebSocket auto-response so an idle fleet never wakes the DO.
HB_FRAME = '{"t":"hb"}'


class E2E(object):
    """Protocol v1e payload encryption (docs/RDD-v2.md): data/cmd values are
    base64( nonce(12) || AES-256-GCM ciphertext ) under a per-session key
    HKDF-SHA256(fleet key, salt=sid, info=b"popfleet-v2")."""

    def __init__(self, fleet_key, sid):
        # Single-block HKDF-SHA256: extract then one expand round (32 bytes).
        prk = hmac.new(sid.encode(), fleet_key, hashlib.sha256).digest()
        okm = hmac.new(prk, b"popfleet-v2\x01", hashlib.sha256).digest()
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM
        self.aead = AESGCM(okm)

    def seal(self, data):
        nonce = os.urandom(12)
        return base64.b64encode(nonce + self.aead.encrypt(nonce, data, None)).decode()

    def open(self, wire):
        """Plaintext bytes, or None for any forged/tampered/wrong-key frame."""
        try:
            raw = base64.b64decode(wire)
            return self.aead.decrypt(raw[:12], raw[12:], None)
        except Exception:
            return None


def fleet_key_from_env():
    raw = os.environ.get("POPFLEET_E2E_KEY", "").strip()
    if not raw:
        return None
    try:
        key = base64.b64decode(raw, validate=True)
    except Exception:
        raise SystemExit("popfleet: POPFLEET_E2E_KEY is not valid base64")
    if len(key) != 32:
        raise SystemExit("popfleet: POPFLEET_E2E_KEY must be 32 bytes "
                         "(generate with: openssl rand -base64 32)")
    try:
        from cryptography.hazmat.primitives.ciphers.aead import AESGCM  # noqa: F401
    except ImportError:
        raise SystemExit("popfleet: POPFLEET_E2E_KEY is set but the "
                         "'cryptography' package is missing: pip install cryptography")
    return key


def log(msg):
    sys.stderr.write("popfleet: %s\n" % msg)
    sys.stderr.flush()


def require(name):
    value = os.environ.get(name, "").strip()
    if not value:
        raise SystemExit(
            "popfleet: %s is not set. Config is env-only: secrets passed as "
            "command-line flags are visible to every user in `ps`." % name)
    return value


def ws_url(raw):
    """Broker base URL -> the agent endpoint, exactly as the Go agent does it.

    https://fleet.example -> wss://fleet.example/ws/agent
    """
    parts = urlsplit(raw)
    scheme = {"http": "ws", "https": "wss", "ws": "ws", "wss": "wss"}.get(parts.scheme)
    if scheme is None or not parts.netloc:
        raise SystemExit("popfleet: bad POPFLEET_URL %r "
                         "(want http/https/ws/wss + host)" % raw)
    return urlunsplit((scheme, parts.netloc,
                       parts.path.rstrip("/") + "/ws/agent", "", ""))


class Pty(object):
    """One login shell (or `cmd`) on its own pseudo-terminal."""

    def __init__(self, cmd):
        self.pid, self.fd = pty.fork()
        if self.pid == 0:                       # child: _exec never returns
            self._exec(cmd)
        os.set_blocking(self.fd, False)

    @staticmethod
    def _exec(cmd):
        shell = os.environ.get("SHELL") or "/bin/sh"
        os.environ.setdefault("TERM", "xterm-256color")
        for secret in ("POPFLEET_TOKEN", "POPFLEET_URL", "POPFLEET_E2E_KEY"):
            os.environ.pop(secret, None)        # the shell has no use for these
        try:
            if cmd:
                os.execvp(shell, [shell, "-c", cmd])
            # leading '-' is the ancient convention for "this is a login shell"
            os.execvp(shell, ["-" + os.path.basename(shell)])
        except OSError:
            pass
        os._exit(127)

    def read(self):
        """Bytes from the shell; b"" once it is gone; None if not ready yet."""
        try:
            return os.read(self.fd, READ_SIZE)
        except OSError as e:
            if e.errno == errno.EAGAIN:
                return None
            return b""                          # EIO on Linux == pty closed

    def write(self, data):
        try:
            os.write(self.fd, data)
        except OSError:
            pass                                # shell died mid-keystroke

    def resize(self, cols, rows):
        try:
            fcntl.ioctl(self.fd, termios.TIOCSWINSZ,
                        struct.pack("HHHH", rows, cols, 0, 0))
        except OSError:
            pass

    def kill(self):
        """Kill the shell and reap it. Returns its exit code."""
        try:
            os.killpg(self.pid, signal.SIGKILL)  # pty.fork() made it a session leader
        except OSError:
            pass                                 # already exited
        try:
            os.close(self.fd)
        except OSError:
            pass
        # ponytail: blocking waitpid. The child is already signalled or already
        # dead, so this returns at once; no reaper thread earns its keep here.
        try:
            _, status = os.waitpid(self.pid, 0)
        except OSError:
            return 0
        code = os.waitstatus_to_exitcode(status)
        return code if code >= 0 else 128 - code  # killed by signal N -> 128+N


class Agent(object):
    """One broker connection and the PTYs that live and die with it."""

    def __init__(self, ws, loop, fleet_key=None):
        self.ws = ws
        self.loop = loop
        self.fleet_key = fleet_key      # None = plaintext v1 (LAN broker)
        self.ptys = {}                  # sid -> Pty
        self.enc = {}                   # sid -> E2E (only when fleet_key set)
        self.obuf = {}                  # sid -> bytearray, coalesced PTY output
        self.oflush = {}                # sid -> TimerHandle for the 16 ms flush
        self.outbox = asyncio.Queue()   # frames waiting for the socket
        self.enrolled = False

    def send(self, **frame):
        self.outbox.put_nowait(json.dumps(frame))

    async def pump_out(self):
        while True:
            await self.ws.send(await self.outbox.get())

    async def heartbeat(self):
        while True:
            await asyncio.sleep(HEARTBEAT_S)
            self.outbox.put_nowait(HB_FRAME)

    async def pump_in(self):
        async for raw in self.ws:
            try:
                frame = json.loads(raw)
            except ValueError:
                continue
            self.handle(frame)

    def handle(self, m):
        tag = m.get("t")
        sid = m.get("sid")
        if tag == "hello_ok":
            self.enrolled = True
            log("enrolled as machine %s" % m.get("id"))
        elif tag == "open":
            self.open(sid, m.get("cmd"))
        elif tag == "in":
            p = self.ptys.get(sid)                  # unknown sid: ignored
            if p is None:
                pass
            elif sid in self.enc:
                data = self.enc[sid].open(m.get("data", ""))
                if data is None:                    # GCM fail: kill, never render
                    self.e2e_kill(sid, "e2e decrypt failed (tampered frame)")
                else:
                    p.write(data)
            else:
                p.write(base64.b64decode(m.get("data", "")))
        elif tag == "resize":
            p = self.ptys.get(sid)
            if p is not None:
                p.resize(int(m.get("c", 80)), int(m.get("r", 24)))
        elif tag == "close":
            self.reap(sid, notify=True)             # killed is still exited
        # unknown tags are ignored on purpose (forward compatibility)

    def open(self, sid, cmd):
        if not sid or sid in self.ptys:
            return                                  # duplicate open: ignored
        enc = None
        if self.fleet_key is not None:              # protocol v1e: cmd is ciphertext
            enc = E2E(self.fleet_key, sid)
            if cmd:
                plain = enc.open(cmd)
                if plain is None:                   # never run what we cannot authenticate
                    self.send(t="err", sid=sid,
                              msg="e2e decrypt failed (key mismatch or tampered cmd)")
                    return
                cmd = plain.decode("utf-8", "replace")
        try:
            p = Pty(cmd)
        except OSError as e:
            log("open %s failed: %s" % (sid, e))
            self.send(t="exit", sid=sid, code=127)
            return
        self.ptys[sid] = p
        if enc is not None:
            self.enc[sid] = enc
        self.loop.add_reader(p.fd, self.drain, sid)

    def drain(self, sid):
        p = self.ptys.get(sid)
        if p is None:
            return
        data = p.read()
        if data is None:
            return
        if data:
            # Coalesce: a `yes` firehose as frame-per-read is a per-frame cost
            # through the Worker relay; batch 16 ms / 4 KiB, whichever first.
            buf = self.obuf.setdefault(sid, bytearray())
            buf.extend(data)
            if len(buf) >= FLUSH_BYTES:
                self.flush(sid)
            elif sid not in self.oflush:
                self.oflush[sid] = self.loop.call_later(FLUSH_S, self.flush, sid)
        else:
            self.reap(sid, notify=True)             # EOF: the shell exited

    def flush(self, sid):
        timer = self.oflush.pop(sid, None)
        if timer is not None:
            timer.cancel()
        buf = self.obuf.pop(sid, None)
        if not buf:
            return
        data = bytes(buf)
        enc = self.enc.get(sid)
        wire = enc.seal(data) if enc else base64.b64encode(data).decode()
        self.send(t="out", sid=sid, data=wire)

    def e2e_kill(self, sid, why):
        """GCM auth failure: report err, kill the PTY, render nothing."""
        log("%s: %s" % (sid, why))
        self.obuf.pop(sid, None)                    # pending output dies with it
        self.send(t="err", sid=sid, msg=why)
        self.reap(sid, notify=False)

    def reap(self, sid, notify):
        p = self.ptys.pop(sid, None)
        if p is None:
            return
        self.flush(sid)                             # last buffered bytes first
        self.enc.pop(sid, None)
        self.loop.remove_reader(p.fd)               # before the fd is closed
        code = p.kill()
        if notify:
            self.send(t="exit", sid=sid, code=code)

    def shutdown(self):
        for sid in list(self.ptys):
            self.reap(sid, notify=False)


async def run_connection(url, token, name, stop, fleet_key):
    """One connection, start to finish. Returns True if we ever enrolled."""
    agent = None
    try:
        async with websockets.connect(url, open_timeout=15, close_timeout=5) as ws:
            agent = Agent(ws, asyncio.get_running_loop(), fleet_key)
            hello = {"t": "hello", "token": token, "name": name, "ver": VER}
            if fleet_key is not None:
                hello["e2e"] = True     # protocol v1e: payloads are ciphertext
            await ws.send(json.dumps(hello))
            tasks = [asyncio.ensure_future(co) for co in (
                agent.pump_in(), agent.pump_out(), agent.heartbeat(), stop.wait())]
            done, pending = await asyncio.wait(
                tasks, return_when=asyncio.FIRST_COMPLETED)
            for t in pending:
                t.cancel()
            for t in done:
                if t.exception() is not None:
                    log("socket lost: %s" % t.exception())
    except Exception as e:      # DNS, refused, TLS, bad handshake: all retryable
        log("connect failed: %s" % e)
    finally:
        if agent is not None:
            agent.shutdown()    # no PTY of ours outlives its socket
    return agent is not None and agent.enrolled


async def serve(url, token, name, fleet_key):
    stop = asyncio.Event()
    loop = asyncio.get_running_loop()
    for sig in (signal.SIGINT, signal.SIGTERM):
        try:
            loop.add_signal_handler(sig, stop.set)
        except (NotImplementedError, RuntimeError, ValueError):
            pass                # embedded in a worker thread: the host owns signals
    delay = BACKOFF_START_S
    while not stop.is_set():
        if await run_connection(url, token, name, stop, fleet_key):
            delay = BACKOFF_START_S     # a connection that enrolled was healthy
        if stop.is_set():
            break
        pause = delay * random.uniform(0.8, 1.2)        # +/-20% jitter
        log("reconnecting in %.1fs" % pause)
        try:
            await asyncio.wait_for(stop.wait(), timeout=pause)
        except asyncio.TimeoutError:
            pass                # slept the full backoff; go round again
        delay = min(delay * 2, BACKOFF_MAX_S)
    log("stopped")


def main():
    url = ws_url(require("POPFLEET_URL"))
    token = require("POPFLEET_TOKEN")
    name = os.environ.get("POPFLEET_NAME", "").strip() or socket.gethostname()
    fleet_key = fleet_key_from_env()
    log("agent %s -> %s as %s (e2e %s)" %
        (VER, url, name, "on" if fleet_key else "off"))
    asyncio.run(serve(url, token, name, fleet_key))


if __name__ == "__main__":
    main()
